File: scripts/patchers/kernel_jb_patch_load_dylinker.py
"""Mixin: KernelJBPatchLoadDylinkerMixin."""


class KernelJBPatchLoadDylinkerMixin:
    def patch_load_dylinker(self):
        """Bypass load_dylinker policy gate in the dyld path.

        Strict selector:
        1. Anchor function by '/usr/lib/dyld' string reference.
        2. Inside that function, find BL <check>; CBZ W0, <allow>.
        3. Replace BL with unconditional B to <allow>.
        """
        self._log("\n[JB] _load_dylinker: skip dyld policy check")

        # Try symbol first
        foff = self._resolve_symbol("_load_dylinker")
        if foff >= 0:
            func_end = self._find_func_end(foff, 0x2000)
            result = self._find_bl_cbz_gate(foff, func_end)
            if result:
                bl_off, allow_target = result
                b_bytes = self._encode_b(bl_off, allow_target)
                if b_bytes:
                    self.emit(
                        bl_off,
                        b_bytes,
                        f"b #0x{allow_target - bl_off:X} [_load_dylinker]",
                    )
                    return True

        # Fallback: strict dyld-anchor function profile.
        str_off = self.find_string(b"/usr/lib/dyld")
        if str_off < 0:
            self._log("  [-] '/usr/lib/dyld' string not found")
            return False

        kstart, kend = self._get_kernel_text_range()
        refs = self.find_string_refs(str_off, kstart, kend)
        if not refs:
            refs = self.find_string_refs(str_off)
        if not refs:
            self._log("  [-] no code refs to '/usr/lib/dyld'")
            return False

        for adrp_off, _, _ in refs:
            func_start = self.find_function_start(adrp_off)
            if func_start < 0:
                continue
            func_end = self._find_func_end(func_start, 0x1200)
            result = self._find_bl_cbz_gate(func_start, func_end)
            if not result:
                continue
            bl_off, allow_target = result
            b_bytes = self._encode_b(bl_off, allow_target)
            if not b_bytes:
                continue
            self._log(
                f"  [+] dyld anchor func at 0x{func_start:X}, "
                f"patch BL at 0x{bl_off:X}"
            )
            self.emit(
                bl_off,
                b_bytes,
                f"b #0x{allow_target - bl_off:X} [_load_dylinker policy bypass]",
            )
            return True

        self._log("  [-] dyld policy gate not found")
        return False

    def _find_bl_cbz_gate(self, start, end):
        """Find BL <check>; CBZ W0,<allow>; MOV W0,#2 gate and return (bl_off, allow_target)."""
        for off in range(start, end - 8, 4):
            d0 = self._disas_at(off)
            d1 = self._disas_at(off + 4)
            d2 = self._disas_at(off + 8)
            if not d0 or not d1:
                continue
            i0 = d0[0]
            i1 = d1[0]
            if i0.mnemonic != "bl" or i1.mnemonic != "cbz":
                continue
            if not i1.op_str.startswith("w0, "):
                continue
            if len(i1.operands) < 2:
                continue
            allow_target = i1.operands[-1].imm

            # Keep selector strict: deny path usually sets errno=2 right after CBZ.
            if d2 and d2[0].mnemonic == "mov" and d2[0].op_str.startswith("w0, #2"):
                return off, allow_target
        return None

File: sources/FirmwarePatcher/Kernel/JBPatches/KernelJBPatchLoadDylinker.swift
// KernelJBPatchLoadDylinker.swift — JB: bypass load_dylinker policy gate in the dyld path.
//
// Historical note: derived from the legacy Python firmware patcher during the Swift migration.
//
// Reveal: string-anchor "/usr/lib/dyld" → kernel-text function containing the ref →
//   inside that function: BL <check>; CBZ W0, <allow>; MOV W0, #2 (deny path).
// Patch: replace BL with unconditional B to <allow>, skipping the policy check.

import Foundation

extension KernelJBPatcher {
    /// Bypass the load_dylinker policy gate in the dyld path.
    @discardableResult
    func patchLoadDylinker() -> Bool {
        log("\n[JB] _load_dylinker: skip dyld policy check")

        guard let strOff = buffer.findString("/usr/lib/dyld") else {
            log("  [-] '/usr/lib/dyld' string not found")
            return false
        }

        let refs = findStringRefs(strOff)
        guard !refs.isEmpty else {
            log("  [-] no kernel-text code refs to '/usr/lib/dyld'")
            return false
        }

        for (adrpOff, _) in refs {
            guard jbIsInCodeRange(adrpOff) else { continue }
            guard let funcStart = findFunctionStart(adrpOff) else { continue }
            let funcEnd = findFuncEnd(funcStart, maxSize: 0x1200)

            guard let (blOff, allowTarget) = findBlCbzGate(funcStart: funcStart, funcEnd: funcEnd) else {
                continue
            }

            guard let bBytes = ARM64Encoder.encodeB(from: blOff, to: allowTarget) else { continue }

            log("  [+] dyld anchor func at 0x\(String(format: "%X", funcStart)), patch BL at 0x\(String(format: "%X", blOff))")
            emit(blOff, bBytes,
                 patchID: "jb.load_dylinker.policy_bypass",
                 virtualAddress: fileOffsetToVA(blOff),
                 description: "b #0x\(String(format: "%X", allowTarget - blOff)) [_load_dylinker policy bypass]")
            return true
        }

        log("  [-] dyld policy gate not found in dyld-anchored function")
        return false
    }

    // MARK: - Private helpers

    /// Scan [funcStart, funcEnd) for `BL <check> ; CBZ W0, <allow> ; MOV W0, #2`.
    /// Returns (blOff, allowTarget) on success.
    private func findBlCbzGate(funcStart: Int, funcEnd: Int) -> (blOff: Int, allowTarget: Int)? {
        var off = funcStart
        while off + 12 <= funcEnd {
            defer { off += 4 }

            let insns0 = disasm.disassemble(in: buffer.data, at: off, count: 1)
            guard let i0 = insns0.first, i0.mnemonic == "bl" else { continue }

            let insns1 = disasm.disassemble(in: buffer.data, at: off + 4, count: 1)
            guard let i1 = insns1.first, i1.mnemonic == "cbz" else { continue }
            guard i1.operandString.hasPrefix("w0, ") else { continue }

            guard let detail1 = i1.aarch64, detail1.operands.count >= 2 else { continue }
            let allowTarget = Int(detail1.operands.last!.imm)

            // Selector: deny path sets w0 = 2 immediately after CBZ.
            let insns2 = disasm.disassemble(in: buffer.data, at: off + 8, count: 1)
            if let i2 = insns2.first,
               i2.mnemonic == "mov",
               i2.operandString.hasPrefix("w0, #2")
            {
                return (off, allowTarget)
            }
        }
        return nil
    }
}
